crates/ruma-events/src/state_key.rs

```diff
@@ -38,3 +39,208 @@ impl Serialize for EmptyStateKey {
         serializer.serialize_str("")
     }
 }
+
+/// A type that can be used as the `state_key` for call member state events.
+/// Those state keys can be a combination of UserId and DeviceId.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[allow(clippy::exhaustive_structs)]
+pub struct CallMemberStateKey {
+    key: CallMemberStateKeyEnum,
+    raw: Box<str>,
+}
+
+impl CallMemberStateKey {
+    /// Constructs a new CallMemberStateKey there are three possible formats:
+    /// - "_{UserId}_{DeviceId}" example: "_@test:user.org_DEVICE". `device_id`: Some`, `underscore:
+    ///   true`
+    /// - "{UserId}_{DeviceId}" example: "@test:user.org_DEVICE". `device_id`: Some`, `underscore:
+    ///   false`
+    /// - "{UserId}" example example: "@test:user.org". `device_id`: None`, underscore is ignored:
+    ///   `underscore: false|true`
+    ///
+    /// Dependent on the parameters the correct CallMemberStateKey will be constructed.
+    pub fn new(user_id: OwnedUserId, device_id: Option<OwnedDeviceId>, underscore: bool) -> Self {
+        CallMemberStateKeyEnum::new(user_id, device_id, underscore).into()
+    }
+
+    /// Returns the user id in this state key.
+    /// (This is a cheap operations. The id is already type checked on initialization. And does only
+    /// returns a reference to an existing OwnedUserId.)
+    pub fn user_id(&self) -> &UserId {
+        match &self.key {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(u, _) => u,
+            CallMemberStateKeyEnum::UserDevice(u, _) => u,
+            CallMemberStateKeyEnum::User(u) => u,
+        }
+        .as_ref()
+    }
+
+    /// Returns the device id in this state key (if available)
+    /// (This is a cheap operations. The id is already type checked on initialization. And does only
+    /// returns a reference to an existing OwnedDeviceId.)
+    pub fn device_id(&self) -> Option<&DeviceId> {
+        match &self.key {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(_, d) => Some(d),
+            CallMemberStateKeyEnum::UserDevice(_, d) => Some(d),
+            CallMemberStateKeyEnum::User(_) => None,
+        }
+        .map(|d| d.as_ref())
+    }
+}
+
+impl AsRef<str> for CallMemberStateKey {
+    fn as_ref(&self) -> &str {
+        &self.raw
+    }
+}
+
+impl From<CallMemberStateKeyEnum> for CallMemberStateKey {
+    fn from(value: CallMemberStateKeyEnum) -> Self {
+        let raw: Box<str> = match &value {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(u, d) => format!("_{}_{}", u, d).into(),
+            CallMemberStateKeyEnum::UserDevice(u, d) => format!("{}_{}", u, d).into(),
+            CallMemberStateKeyEnum::User(u) => u.to_string().into(),
+        };
```

Comment:
As I said in a previous comment, we should implement `std::fmt::Display` for `CallMemberStateKeyEnum` and call it here.